src/number_theory/number_field/canonical_embedding.lean

```diff
@@ -0,0 +1,171 @@
+/-
+Copyright (c) 2022 Xavier Roblot. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Xavier Roblot
+-/
+import number_theory.number_field.embeddings
+
+/-!
+# Canonical embedding of a number field
+
+The canonical embedding of a number field `K` of signature `(r₁, r₂)` is the ring homomorphism
+`K →+* ℝ^r₁ × ℂ^r₂` that sends `x ∈ K` to `(φ_₁(x),...,φ_r₁(x)) × (ψ_₁(x),..., ψ_r₂(x))` where
+`φ_₁,...,φ_r₁` are its real embeddings and `ψ_₁,..., ψ_r₂` are its complex embeddings (up to
+complex conjugation).
+
+## Main definitions and results
+
+* `number_field.canonical_embedding.ring_of_integers.inter_ball_finite`: the intersection of the
+image of the ring of integers by the canonical embedding and any ball centered at `0` of finite
+radius is finite.
+
+## Tags
+
+number field, infinite places
+-/
+
+noncomputable theory
+
+open function finite_dimensional finset fintype number_field number_field.infinite_place metric
+module
+open_locale classical number_field
+
+variables (K : Type*) [field K]
+
+namespace number_field.canonical_embedding
+
+/-- The ambiant space `ℝ^r₁ × ℂ^r₂` with `(r₁, r₂)` the signature of `K`. -/
+@[reducible] def space :=
+({w : infinite_place K // is_real w} → ℝ) × ({w : infinite_place K // is_complex w} → ℂ)
+
+instance : comm_ring (space K) := prod.comm_ring
+
+instance : module ℝ (space K) := prod.module
+
+lemma space_rank [number_field K] :
+  finrank ℝ (space K) = finrank ℚ K :=
+begin
+  haveI : module.free ℝ ℂ := infer_instance,
+  rw [module.free.finrank_prod, module.free.finrank_pi, module.free.finrank_pi_fintype,
+    complex.finrank_real_complex, finset.sum_const, finset.card_univ, ← card_real_embeddings,
+    algebra.id.smul_eq_mul, mul_comm, ← card_complex_embeddings, ← number_field.embeddings.card K ℂ,
+    fintype.card_subtype_compl, nat.add_sub_of_le (fintype.card_subtype_le _)],
+end
+
+@[nolint fails_quickly]
```

Comment:
Is there a loop here? In any case I think a comment explaining why we had to disable the linter is a good idea.

Reply:
There is no loop. I'll let it resolve on my laptop and it did resolve but after a long time, say maybe 30mn. I don't know exactly as I let it run while I was having lunch ;)

Reply:
Hmm, we should at least try to understand what's going wrong here.

Reply:
Ok. I'll run it again and keep the trace to try to see what's happening there.

Reply:
So I was a bit impatient since in fact it takes only about 5mn on my laptop for the instance to fail. That is still too long for the linter though. The trace file is 13MB and has about 215000 lines. I had a quick look and did not see anything suspicious (but I am not an expert). I can post it somewhere if you want to see it (it's only 1MB once zipped). In any case, I think that is a good argument for going back to a local definition and removing the two instances.

Reply:
So it now works as expected? I mean, is it reasonably fast?

Reply:
I removed the instances so there is no problem anymore.